docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,201 @@
+
+  publicKey: LS0tLS1CRUdJTiBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCi4uLgotLS0tLUVORCBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCg== # user's PGP public key.
+
+  clientIP: # external ip of the user
+    ipv4: 1.2.3.4
+    # ipv6: ::1
+
+status:
+  # the following fields are managed by the controller in the seed and synced by gardenlet
+  bastionIP: 1.2.3.5
+  state: Ready
+  id_rsa.enc: LS0tLS1CRUdJTiBQR1AgTUVTU0FHRS0tLS0tCi4uLgotLS0tLUVORCBQR1AgTUVTU0FHRS0tLS0tCg== # SSH private key, enrypted with spec.publicKey
+  id_rsa.pub: c3NoLXJzYSAuLi4K
+
+  # the following fields are only set by the mutating webhook
+  expirationDate: "2021-03-19T12:58:00Z" # extended on each keepalive
+```
+
+`Bastion` custom resource in the seed cluster
+```yaml
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Bastion
+metadata:
+  name: cli-abcdef
+  namespace: shoot--myproject--mycluster
+  annotations:
+    operations.gardener.cloud/created-by: foo
+    operations.gardener.cloud/last-heartbeat-at: "2021-03-19T11:58:00Z"
+spec:
+  publicKey: LS0tLS1CRUdJTiBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCi4uLgotLS0tLUVORCBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCg== # user's PGP public key.
+
+  clientIP: # external ip of the user
+    ipv4: 1.2.3.4
+    # ipv6: ::1
+
+status:
+  bastionIP: 1.2.3.5
+  state: Ready
+  id_rsa.enc: LS0tLS1CRUdJTiBQR1AgTUVTU0FHRS0tLS0tCi4uLgotLS0tLUVORCBQR1AgTUVTU0FHRS0tLS0tCg== # SSH private key, enrypted with spec.publicKey
+  id_rsa.pub: c3NoLXJzYSAuLi4K
+
+  expirationDate: "2021-03-19T12:58:00Z"
+```
+
+## SSH Key Pair Rotation
+Currently, the SSH key pair for the shoot nodes are created once during shoot cluster creation. These key pairs should be rotated on a regular basis.
+
+### Rotation Proposal
+- `gardeneruser` original user data [component](https://github.com/gardener/gardener/tree/master/pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/gardeneruser):
+    - The `gardeneruser` [create script](https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/gardeneruser/templates/scripts/create.tpl.sh) should be changed into a reconcile script script, and renamed accordingly. It needs to be adapted so that the `authorized_keys` file will be updated / overwritten with the current SSH public key from the cloud-config user data.
+- Rotation trigger:
+    - once in the maintenance time window
+    - On demand, by annotating the shoot with `gardener.cloud/operation: rotate-ssh-keypair`
+- On rotation trigger:
+    - `gardenlet`
+        - Prerequisite of SSH key pair rotation: all nodes of all the worker pools have successfully applied the desired version of their cloud-config user data
+        - Creates secret `ssh-keypair.old` with the content of `ssh-keypair` in the seed-shoot namespace. The old private key can be used by clients as fallback, in case the new SSH public key is not yet applied on the node
```

Comment:
Will both keys be valid from some time? It would be good, if the old key is dropped only with the next maintenance time window. Then we have no issues if that happens during something like an operator shift.

Reply:
For one, not even `gardenctl` can do that atomically safe, if a key gets rotated right under its nose. Then, the same problem applies to the nodes themselves, that don't update atomically (first the keys are replaced, then the nodes see the new key). Maybe these are theoretical issues, but that's what can happen in complex software systems like ours/these.

On top, nobody forces the use of `gardenctl` and if I get my bastion and credentials for 8h or whatever and use my own `ssh` clients, I expect them to work, also if I trigger a maintenance operation or it happens while I analyse an issue.

In general, you find that pattern (both are valid for some time) in many places, the old keys are always valid in the transition period until the new keys are rolled out everywhere and in this application of this concept everywhere includes the operator, so we shall count that in and the easiest way is to just kick it out with the next maintenance.

Reply:
@vlerenc I have updated the text to keep the old key until the next rotation trigger and to apply the old and current key on the node, see diff here https://github.com/gardener/gardener/pull/3802/commits/8c5fbd01cb54e0bd599756cb3642346ff9f5a4d6